hypershift-operator/controllers/nodepool/openstack/openstack.go

```diff
@@ -72,3 +82,85 @@ func MachineTemplateSpec(hcluster *hyperv1.HostedCluster, nodePool *hyperv1.Node
 	}
 	return openStackMachineTemplate, nil
 }
+
+func getOpenStackClusterForHostedCluster(ctx context.Context, c client.Client, hcluster *hyperv1.HostedCluster, controlPlaneNamespace string) (capiopenstackv1beta1.OpenStackCluster, error) {
+	cluster := capiopenstackv1beta1.OpenStackCluster{}
+
+	if err := c.Get(ctx, types.NamespacedName{Namespace: controlPlaneNamespace, Name: hcluster.Name}, &cluster); err != nil {
+		return cluster, fmt.Errorf("failed to list Machines: %w", err)
+	}
+
+	return cluster, nil
+}
+
+func ReconcileOpenStackImageCR(ctx context.Context, client client.Client, createOrUpdate upsert.CreateOrUpdateFN, hcluster *hyperv1.HostedCluster, release *releaseinfo.ReleaseImage, nodePool *hyperv1.NodePool) error {
+	releaseVersion, err := releaseinfo.OpenStackReleaseImage(release)
+	if err != nil {
+		return err
+	}
+	controlPlaneNamespace := fmt.Sprintf("%s-%s", nodePool.Namespace, strings.ReplaceAll(nodePool.Spec.ClusterName, ".", "-"))
+	openstackCluster, err := getOpenStackClusterForHostedCluster(ctx, client, hcluster, controlPlaneNamespace)
+	if err != nil {
+		return err
+	}
+	openStackImage := orc.Image{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rhcos-" + releaseVersion + "-" + hcluster.Name,
+			Namespace: controlPlaneNamespace,
+			// TODO: add proper cleanup in CAPI resources cleanup
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: openstackCluster.APIVersion,
+					Kind:       openstackCluster.Kind,
+					Name:       openstackCluster.Name,
+					UID:        openstackCluster.UID,
+				},
+			},
+		},
+		Spec: orc.ImageSpec{},
+	}
+
+	if _, err := createOrUpdate(ctx, client, &openStackImage, func() error {
+		err := reconcileOpenStackImageSpec(hcluster, &openStackImage.Spec, release)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func reconcileOpenStackImageSpec(hcluster *hyperv1.HostedCluster, openStackImageSpec *orc.ImageSpec, release *releaseinfo.ReleaseImage) error {
+	imageURL, imageHash, err := releaseinfo.UnsupportedOpenstackDefaultImage(release)
+	if err != nil {
+		return fmt.Errorf("failed to lookup RHCOS image: %w", err)
+	}
+
+	openStackImageSpec.CloudCredentialsRef = orc.CloudCredentialsReference{
+		SecretName: hcluster.Spec.Platform.OpenStack.IdentityRef.Name,
+		CloudName:  hcluster.Spec.Platform.OpenStack.IdentityRef.CloudName,
+	}
+	releaseVersion, err := releaseinfo.OpenStackReleaseImage(release)
+	if err != nil {
+		return err
+	}
+
+	openStackImageSpec.Resource = &orc.ImageResourceSpec{
+		Name: "rhcos-" + releaseVersion + "-" + hcluster.Name,
+		Content: &orc.ImageContent{
+			DiskFormat: "qcow2",
+			Download: &orc.ImageContentSourceDownload{
+				URL:        imageURL,
+				Decompress: ptr.To(orc.ImageCompressionGZ),
+				Hash: &orc.ImageHash{
+					Algorithm: "sha256",
+					Value:     imageHash,
+				},
+			},
+		},
+	}
```

Comment:
In order to be efficient, in CI we will need to specify `managedOptions.onDelete` to `detach` so the RHCOS image will be created once in Glance and re-used across multiple Hosted Clusters. The CI cleanup jobs will eventually take care of deleting the image but the Hypershift CI wouldn't bother doing it.

This is a CI use case. As a customer, the image would always be deleted when a HostedCluster is being deleted.

cc @mdbooth

Reply:
done